pkg/manifest/firstboot.go

```diff
@@ -0,0 +1,168 @@
+package manifest
+
+import (
+	"fmt"
+	"html/template"
+	"io/fs"
+	"strings"
+
+	"github.com/osbuild/images/internal/common"
+	"github.com/osbuild/images/pkg/customizations/firstboot"
+	"github.com/osbuild/images/pkg/customizations/fsnode"
+	"github.com/osbuild/images/pkg/osbuild"
+)
+
+// checkName prevents path traversal
+func checkName(str string) error {
+	if str == "" {
+		return fmt.Errorf("name cannot be empty")
+	}
+
+	for _, r := range str {
+		if !(('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || ('0' <= r && r <= '9') || r == '-' || r == '_') {
+			return fmt.Errorf("name can only contain alphanumeric characters, dashes, and underscores")
+		}
+	}
+
+	return nil
+}
+
+var tmplFirstbootAAP = `#!/usr/bin/bash
+curl -s -i --data "host_config_key={{ .HostConfigKey }}" {{ .URL -}}
+`
+
+func renderFirstboot(tmplStr string, data any) (string, error) {
+	tmpl, err := template.New("firstboot-unit").Parse(tmplStr)
+	if err != nil {
+		return "", fmt.Errorf("error parsing firstboot unit template: %w", err)
+	}
+
+	var result strings.Builder
+	err = tmpl.Execute(&result, data)
+	if err != nil {
+		return "", fmt.Errorf("error rendering firstboot unit: %w", err)
+	}
+
+	return result.String(), nil
+}
+
+func firstbootFileNodes(fbo *firstboot.FirstbootOptions) ([]string, []*fsnode.File, *osbuild.SystemdUnitCreateStageOptions, error) {
+	if fbo == nil {
+		return nil, nil, nil, nil
+	}
+
+	var certs []string
+	var files []*fsnode.File
+	var executables []string
+
+	var ci int
+	eachCustom := func(opt firstboot.CustomFirstbootOptions) error {
+		// keep the naming convention consistent with the existing "osbuild-first-boot"
+		name := fmt.Sprintf("osbuild-first-%s", opt.Name)
+		if checkName(opt.Name) != nil {
+			ci++
+			name = fmt.Sprintf("osbuild-first-custom-%d", ci)
+		}
+
+		// create the executable
+		exec := fmt.Sprintf("/usr/local/bin/%s", name)
+		f, err := fsnode.NewFile(exec, common.ToPtr(fs.FileMode(0770)), "root", "root", []byte(opt.Contents))
+		if err != nil {
+			return fmt.Errorf("error creating firstboot file node: %w", err)
+		}
+		files = append(files, f)
+
+		// prepare data for the systemd unit
+		if opt.IgnoreFailure {
+			exec = "-" + exec
+		}
+		executables = append(executables, exec)
```

Comment:
This bit is repeated in each `each*` function, lift it to a helper?

Reply:
It is really on the edge of 3-4 lines where I deem it acceptable and it is very local stuff. I think perhaps inner function might be ok let me try.